cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -0,0 +1,353 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "parquet_gpu.hpp"
+#include <cudf/detail/utilities/integer_utils.hpp>
+
+namespace cudf {
+namespace io {
+namespace parquet {
+namespace gpu {
+
+// TODO: consider if these should be template parameters to rle_stream
+constexpr int num_rle_stream_decode_threads = 512;
+constexpr int num_rle_stream_decode_warps   = (num_rle_stream_decode_threads / 32) - 1;
+constexpr int run_buffer_size               = (num_rle_stream_decode_warps * 2);
+constexpr int rolling_run_index(int index) { return index % run_buffer_size; }
+
+/**
+ * @brief Read a 32-bit varint integer
+ *
+ * @param[in,out] cur The current data position, updated after the read
+ * @param[in] end The end data position
+ *
+ * @return The 32-bit value read
+ */
+inline __device__ uint32_t get_vlq32(const uint8_t*& cur, const uint8_t* end)
+{
+  uint32_t v = *cur++;
+  if (v >= 0x80 && cur < end) {
+    v = (v & 0x7f) | ((*cur++) << 7);
+    if (v >= (0x80 << 7) && cur < end) {
+      v = (v & ((0x7f << 7) | 0x7f)) | ((*cur++) << 14);
+      if (v >= (0x80 << 14) && cur < end) {
+        v = (v & ((0x7f << 14) | (0x7f << 7) | 0x7f)) | ((*cur++) << 21);
+        if (v >= (0x80 << 21) && cur < end) {
+          v = (v & ((0x7f << 21) | (0x7f << 14) | (0x7f << 7) | 0x7f)) | ((*cur++) << 28);
+        }
+      }
+    }
+  }
+  return v;
+}
+
+// an individual batch. processed by a warp.
+// batches should be in shared memory.
+template <typename level_t>
+struct rle_batch {
+  uint8_t const* run_start;  // start of the run we are part of
+  int run_offset;            // value offset of this batch from the start of the run
+  level_t* output;
+  int level_run;
+  int size;
+
+  __device__ inline void decode(uint8_t const* const end, int level_bits, int lane, int warp_id)
+  {
+    int output_pos = 0;
+    int remain     = size;
+
+    // for bitpacked/literal runs, total size is always a multiple of 8. so we need to take care if
+    // we are not starting/ending exactly on a run boundary
+    uint8_t const* cur;
+    if (level_run & 1) {
+      int const effective_offset = cudf::util::round_down_safe(run_offset, 8);
+      int const lead_values      = (run_offset - effective_offset);
+      output_pos -= lead_values;
+      remain += lead_values;
+      cur = run_start + ((effective_offset >> 3) * level_bits);
+    }
+
+    // if this is a repeated run, compute the repeated value
+    int _level_val;
+    if (!(level_run & 1)) {
+      _level_val = run_start[0];
+      if (level_bits > 8) { _level_val |= run_start[0] << 8; }
+    }
+
+    // process
+    while (remain > 0) {
+      int batch_len = min(32, remain);
+
+      // if this is a literal run. each thread computes it's own level_val
+      if (level_run & 1) {
+        int const batch_len8 = (batch_len + 7) >> 3;
```

Comment:
```suggestion
        int const batch_len8 = cudf::util::div_rounding_up_unsafe(batch_len, 8);
```

Reply:
Is this variable the number of batches of eight elements?
It's not used in the branch, so it can be declared after.